### ISSUE TITLE
Rust reproducibility issue - Finding the proper fix

### ISSUE THREAD

**Issue by sundeep-kokkonda** [rust-lang/rust#134589]
We are building Rust for a custom target in the Yocto framework. Here we run a test called _reproducibility_ to ensure the rust binaries are identical between the builds.
We do that by building and comparing the rust binaries & libs in two different build directories -
1. _reproducibleA_ &
2. _reproducibleB-extended_

With every rust release we've seen a several issues which makes the rust binaries are not reproducible. The issues seen are with several reasons like -

1. Regressions caused by rust updates (absolute path in proc-macro)
https://github.com/rust-lang/rust/pull/121959/commits/a9a979839bbdfec48c75d618ab0dce8a953589b8

2. Incorrect/missing Rust config settings (_remap-debuginfo_ is not set by default)
https://lists.openembedded.org/g/openembedded-core/message/188940

3. Compiler options and settings - _rustdoc_ failed with _codegen-units_ option setting
https://lists.openembedded.org/g/openembedded-core/message/202540

4. Regressions with Crate updates - CC crate appending a unique hash to the object files
https://github.com/rust-lang/cc-rs/pull/1277
https://github.com/rust-lang/cc-rs/pull/1270

There are a few other failure cases we've analyzed and fixed. This issue is happening for multiple reasons and these failure are resulting in lot of debug efforts and project delays.

This can be avoided by having such a reproducibility test on _rust_ itself.

Any comments or inputs on detecting and fixing this issue during rust development itself? (via a test case or any other ways)

**Comment by workingjubilee** [comment]
@sundeep-kokkonda Yocto's Poky build system modifies the Rust toolchain in several ways to make it more... reproducible... and unfortunately often avoids upstreaming such changes. This means that Yocto avoids receiving feedback on whether such changes are appropriate, or why. This also means that we do not hear about all of your reproducibility concerns, either, only when Yocto thinks that a workaround downstream would be harder. It is unfortunate that you frequently have complications, but unfortunately not surprising.

I would like to acknowledge, for note, that _you_ personally do frequently pursue upstream resolutions. I am merely noting this because I cannot track every patch that Poky maintains on the Rust toolchain to upstream rustc or cargo issues, and a number do not seem to be authored or reviewed by you.

Your proposal sounds fine, but would be most meaningful if Yocto also tried to adopt an upstream-first stance and worked to remove all patches from the Poky build system.

It would be most ideal if there was an attempt to audit for or document what the most common sources of reproducibility issues are. I mean, "build paths embedded in the binary", obviously, but in more detail (common places build paths get introduced, etc.). Somewhere in the rustc-dev-guide, perhaps?

**Comment by jieyouxu** [comment]
To add, a major challenge for the reproducibility problems is that it takes someone (more like, multiple interested contributors) to actually go back and revisit/re-triage the issues, and then also preferrably document the specific places where this is problematic (issues are prefectly fine) and how one might be able to achieve the irreprodubility. When you just say "rustc is not reproducible" or "bootstrap is not reproducible", it's not actionable -- there are *a lot* of places that might be env-sensitive or somehow non-deterministic, and without knowing specifically what, it's just going to be buried in the issue tracker.

**Comment by sundeep-kokkonda** [comment]
@workingjubilee 
> Yocto's Poky build system modifies the Rust toolchain in several ways to make it more... reproducible... and unfortunately often avoids upstreaming such changes.

I can see we duly communicating the pathes in rust code to upstream, I looked into current Yocto rust patches and the changes we made for rust build to work in Yocto environment are reported to upstream.
https://github.com/rust-lang/cargo/issues/13922
https://github.com/rust-lang/cc-rs/pull/1277
https://github.com/madler/zlib/pull/764

> Your proposal sounds fine, but would be most meaningful if Yocto also tried to adopt an upstream-first stance and worked to remove all patches from the Poky build system.

The remaining patches are for testsuit which disables a few tests that are getting failed in Yocto environment (we are currenltly working on fixing those tests).

Anyway, We will work more closely w/ upstream in future by bringing the issues to upstream notice (like test failures... etc)

> what the most common sources of reproducibility issues are.

The repro issues are seen not only with _embedded build path_ but for several other reasons like mentioned in the https://github.com/rust-lang/rust/issues/134589#issue-2753629767 (options, confg, code changes etc). We are open to work with upstream on detecting & fixing it by writing a tests/any other better ways. Looking for your inputs on it.

**Comment by workingjubilee** [comment]
Significant thanks to everyone involved in finally making hardcodepaths.patch go away in this commit https://github.com/yoctoproject/poky/commit/e51f98d5ff41c72590c63acc96da4d8d68d1c8f0#diff-c58dd0952c5853fdc993e22320b91cfac6c7c301cc4ae718aa8a0fb2b5acd479

**Comment by kanavin** [comment]
I've produced a better fix for the rustdoc issue:
https://patchwork.yoctoproject.org/project/oe-core/patch/20250204104205.3423548-1-alex.kanavin@gmail.com/

The key element is adding -Clto=off for rustdoc build only:
```
+        cargo.rustflag("-Clto=off");
```

Why librustdoc is problematic with llvm's lto, but not anything else, remains a mystery.

@sundeep-kokkonda we probably need a separate ticket to track specifically rustdoc? Can you open that please?

**Comment by sundeep-kokkonda** [comment]
> we probably need a separate ticket to track specifically rustdoc? Can you open that please?

@kanavin This issue is created specifically to address all kind of reproducibility issues with rust. The goal is _not_ to fix a single occurrence of reproducibility issue with rust/rustdoc/... but to find a proper fix of all kind of reproducibility issues with rust. So, I do not think a separate issue for _rustdoc_ is not required _or_ you want to track this llvm related bug in rustdoc with other issue?

**Comment by kanavin** [comment]
> So, I do not think a separate issue for _rustdoc_ is not required _or_ you want to track this llvm related bug in rustdoc with other issue?

Yes I do. To get it solved properly we need llvm specialists to look at it, and for that it needs its own ticket.

**Comment by sundeep-kokkonda** [comment]
nondeterministic anon.llvm.<hash> issue - https://github.com/rust-lang/rust/issues/126976

**Comment by windsunil** [comment]
I have a related PR open: #149888. This PR adds a repro-check host tool that performs the core function of building the toolchain twice and comparing outputs, which could serve as the basis for a CI test.

https://github.com/rust-lang/rust/pull/149888
